addons/account/models/account_payment.py

```diff
@@ -186,585 +289,431 @@ def _get_method_codes_needing_bank_account(self):
-            return rec
-        invoices = self.env['account.move'].browse(active_ids)
-
-        # Check all invoices are open
-        if any(invoice.state != 'posted' or invoice.payment_state not in ('not_paid', 'partial') or not invoice.is_invoice() for invoice in invoices):
-            raise UserError(_("You can only register payments for open invoices"))
-        # Check all invoices are inbound or all invoices are outbound
-        outbound_list = [invoice.is_outbound() for invoice in invoices]
-        first_outbound = invoices[0].is_outbound()
-        if any(x != first_outbound for x in outbound_list):
-            raise UserError(_("You can only register at the same time for payment that are all inbound or all outbound"))
-        if any(inv.company_id != invoices[0].company_id for inv in invoices):
-            raise UserError(_("You can only register at the same time for payment that are all from the same company"))
-        # Check the destination account is the same
-        destination_account = invoices.line_ids.filtered(lambda line: line.account_internal_type in ('receivable', 'payable')).mapped('account_id')
-        if len(destination_account) > 1:
-            raise UserError(_('There is more than one receivable/payable account in the concerned invoices. You cannot group payments in that case.'))
-        if 'invoice_ids' not in rec:
-            rec['invoice_ids'] = [(6, 0, invoices.ids)]
-        if 'journal_id' not in rec:
-            rec['journal_id'] = self.env['account.journal'].search([('company_id', '=', self.env.company.id), ('type', 'in', ('bank', 'cash'))], limit=1).id
-        if 'payment_method_id' not in rec:
-            if invoices[0].is_inbound():
-                domain = [('payment_type', '=', 'inbound')]
-            else:
-                domain = [('payment_type', '=', 'outbound')]
-            rec['payment_method_id'] = self.env['account.payment.method'].search(domain, limit=1).id
-        return rec
-
-    @api.depends('invoice_ids', 'journal_id.inbound_payment_method_ids', 'journal_id.outbound_payment_method_ids')
-    def _compute_available_payment_methods(self):
-        for p in self:
-            invoice = p.invoice_ids[:1]
-            if not invoice:
-                p.available_payment_methods = self.env['account.payment.method']
-            elif invoice.is_inbound():
-                p.available_payment_methods = self.journal_id.inbound_payment_method_ids._origin
-            else:
-                p.available_payment_methods = self.journal_id.outbound_payment_method_ids._origin
+            # Update the existing journal items.
+            # If dealing with multiple write-off lines, they are dropped and a new one is generated.
 
-    def _prepare_payment_vals(self, invoices):
-        '''Create the payment values.
+            line_ids_commands = [
+                (1, liquidity_lines.id, line_vals_list[0]),
+                (1, counterpart_lines.id, line_vals_list[1]),
```

Comment:
So you're hacking the position of each line ? :s What guarantee to you have that index 0 and 1 correspond to the lines you want, except you know how it's done directly within the function ? What if this function gets extended ? We should not depend on the order of what it returns.

Reply:
The only way to do that is to make the `_prepare_move_line_default_vals` returning a dictionary...